### PR TITLE
Rewrite of the Logging

### PR DESCRIPTION
**Pull request by Adrian-Weihs**
**Description**
Rewrite of the Logging in AscEmu, the methods have been built to be similar to apach's log4j. The name Logger is a much better name for Logging than Log. Also a log message have now a Severity (Info, Warning, Failure, Fatal) and a MessageType (Trace, Debug, Minor, Major). The configuration for logon and world changed. Some errors which will crash/close the server are now Fatal Messages.

**Todo / Checklist**
- [X] Target is branch develop.
- [X] Detailed description about this pull request.
- [X] Checked AE-Coding standards.

**Tests Performed:** 
- [X] Build AE.
- [X] Server startup.
- [] Log into world.

### CODE REVIEW

**Comment by schnek** [comment]
1. resources.aps wtf?
image
2. LOGGER.debug -> logger.debug i would change case
3. has the work become faster?🤔

**Comment by Adrian-Weihs** [comment]
> 
> 
>     1. resources.aps wtf?
>        image
> 
>     2. LOGGER.debug -> logger.debug i would change case
> 
>     3. has the work become faster?🤔

1. Ups shouldn't pushed it
2. mhh its a define? I come mainly from JAVA so there we write final static variables upper case
3. u mean "faster" performance whise? i dont think so it should be the same but its more clear now

**Comment by Adrian-Weihs** [comment]
ok i messed up

**Comment by schnek** [comment]
> mhh its a define? I come mainly from JAVA so there we write final static variables upper case

yes
`#define logger AscEmu::Logging::Logger::getInstance()`

> u mean "faster" performance whise? i dont think so it should be the same but its more clear now

ok, I'll take a look

**Comment by Zyres** [comment]
Hey Adrian 😘 nice to See something from you, how ist it going?

**Comment by schnek** [comment]
Renamed logger to AElogger.
Renamed LOGGER to logger.

**Comment by Adrian-Weihs** [comment]
> 
> 
> Hey Adrian 😘 nice to See something from you, how ist it going?

Hey :-* I'm fine and u? If u want we can make a discord call tonight.

**Comment by Appled** [comment]
Thank you for your PR 😃 
I would rename "logger" to "sLogger" because most of our Singleton classes use that kind of definition. Personally, I would also rename following functions under Logger class: 'debug' to 'writeDebug', 'trace' to 'writeTrace', 'fatal' to 'writeFatal' etc etc, to keep better clarity what the functions are doing, but I let @Zyres decide on that.

I will do a proper review when the build is fixed 😃

**Comment by schnek** [comment]
logger.debug looks much better than logger.writeDebug it seems to me. The only thing is that the console is now black and white and looks boring.

image
as an option additionally make color...

**Comment by Adrian-Weihs** [comment]
Yeah will rename it to sLogger, but i would say i will stay with debug() etc as method names. I will add a log method where u can set the color

**Review by Appled — CHANGES_REQUESTED** [review]
It's good, just couple codestyle issues.
Also, update the "last edited" on both config files.

File: src/shared/CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +14,8 @@ set(sources
     Database/CreateInterface.cpp
     Database/Database.cpp
     Database/MySQLDatabase.cpp
+	
+	Logging/Logger.cpp
```
[the remark this answers is not in the document]
Comment (Adrian-Weihs) [reply src/shared/CMakeLists.txt]: done

File: src/shared/CMakeLists.txt [hunk 2]
```diff
@@ -75,6 +77,10 @@ set(headers
     Database/MySQLDatabase.h
     Exceptions/Exceptions.hpp
     Exceptions/PlayerExceptions.hpp
+	Logging/ConsoleDefines.hpp
+	Logging/Logger.hpp
+	Logging/MessageType.hpp
+	Logging/Severity.hpp
```
[the remark this answers is not in the document]
Comment (Adrian-Weihs) [reply src/shared/CMakeLists.txt]: done

File: src/shared/LogDefines.hpp [hunk 3]
```diff
@@ -53,4 +53,6 @@ enum LogLevel
     LL_DEBUG     = 2
 };
 
+
+
```
[the remark this answers is not in the document]
Comment (Adrian-Weihs) [reply src/shared/LogDefines.hpp]: done

File: src/shared/Logging/Logger.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    public:
+        Logger(Logger&&) = delete;
+        Logger(Logger const&) = delete;
+        Logger& operator=(Logger&&) = delete;
+        Logger& operator=(Logger const&) = delete;
+
+        static Logger& getInstance();
+
+        void finalize()
```
[the remark this answers is not in the document]
Comment (Adrian-Weihs) [reply src/shared/Logging/Logger.hpp]: done

File: src/shared/Logging/Logger.hpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        void writeFile(FILE* file, const char* msg);
+
+#ifndef _WIN32
+        void setConsoleColor(const char* color);
+#else
+        void setConsoleColor(int color);
+#endif
+
+        void setSeverityConsoleColor(Severity severity);
+
```
[the remark this answers is not in the document]
Comment (Adrian-Weihs) [reply src/shared/Logging/Logger.hpp]: done

**Review by Appled — APPROVED** [review]
👍

**Pull request merged**